src/pydvl/value/shapley/classwise.py

```diff
@@ -0,0 +1,575 @@
+            pbar.refresh()
+            if terminate_exec:
+                break
+
+            for _ in range(n_submitted_jobs - len(futures)):
+                future = executor.submit(
+                    _permutation_montecarlo_classwise_shapley,
+                    u_ref,
+                    truncation=truncation,
+                    done_sample_complements=done_sample_complements,
+                    use_default_scorer_value=use_default_scorer_value,
+                    min_elements_per_label=min_elements_per_label,
+                )
+                futures.add(future)
+
+    result = accumulated_result
+    if normalize_values:
+        result = _normalize_classwise_shapley_values(result, u)
+
+    return result
+
+
+def _permutation_montecarlo_classwise_shapley(
+    u: Utility,
+    *,
+    done_sample_complements: StoppingCriterion = None,
+    truncation: TruncationPolicy,
+    use_default_scorer_value: bool = True,
+    min_elements_per_label: int = 1,
+) -> ValuationResult:
+    """Computes classwise Shapley value using truncated Monte Carlo permutation
+    sampling for the subsets.
+
+    Args:
+        u: Utility object containing model, data, and scoring function. The scoring
+            function should be of type :class:`~pydvl.utils.score.ClassWiseScorer`.
+        done_sample_complements: Function checking whether computation needs to stop.
+            Otherwise, it will resample conditional sets until the stopping criterion is
+            met.
+        truncation: Callable function that decides whether to interrupt processing a
+            permutation and set subsequent marginals to zero.
+        use_default_scorer_value: The first set of indices is the sampled complement
+            set. Unless not otherwise specified, the default scorer value is used for
+            this. If it is set to false, the base score is calculated from the utility.
+        min_elements_per_label: The minimum number of elements for each opposite
+            label.
+
+    Returns:
+        ValuationResult object containing computed data values.
+    """
+    if done_sample_complements is None:
+        done_sample_complements = MaxChecks(1)
+
+    result = ValuationResult.zeros(
+        algorithm="classwise_shapley",
+        indices=u.data.indices,
+        data_names=u.data.data_names,
+    )
+    x_train, y_train = u.data.get_training_data(u.data.indices)
+    unique_labels = np.unique(y_train)
+    scorer = cast(ClasswiseScorer, copy(u.scorer))
+    u.scorer = scorer
+
+    for label in unique_labels:
+        u.scorer.label = label
+        result += _permutation_montecarlo_classwise_shapley_for_label(
+            u,
+            label,
+            done=done_sample_complements,
+            truncation=truncation,
+            use_default_scorer_value=use_default_scorer_value,
+            min_elements_per_label=min_elements_per_label,
+        )
+
+    return result
+
+
+def _check_classwise_shapley_utility(u: Utility):
```

Comment:
A function which is used just once is premature generalization. Move into the body of the calling function.

Reply:
I wouldn't call it premature generalization but more a helper method (although the place might not be the best). But for consistency I removed the function and placed the checking code directly in the function, because the complete (nested) function definition with a docstring makes the method very big.